### PR TITLE
Adding support for uploading the delta from Last Upload in builds

### PR DESCRIPTION
**Pull request by rshunim**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-9754

## Description
In the ticket.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Tests/scripts/collect_tests/collect_tests.py [hunk 1]
```diff
@@ -1110,7 +1143,7 @@ def _collect_single(self, path: Path) -> CollectionResult | None:
             for test in tests)
         )
 
-    def _get_git_diff(self) -> FilesToCollect:
+    def _get_git_diff(self, upload_delta_from_last_upload: bool = False) -> FilesToCollect:
```
Comment (dorschw) [inline Tests/scripts/collect_tests/collect_tests.py:1146]: please document the variable, add method docstrings, as it's a bit surprising here (while necessary)

File: Tests/scripts/infrastructure_tests/test_collect_tests.py [hunk 2]
```diff
@@ -124,11 +124,12 @@ class MockerCases:
...
 
 
-def _test(monkeypatch, case_mocker: CollectTestsMocker, collector_class: Callable,
+def _test(mocker, monkeypatch, case_mocker: CollectTestsMocker, collector_class: Callable,
           expected_tests: Iterable[str], expected_packs: Iterable[str], expected_packs_to_upload: Iterable[str],
           expected_machines: Iterable[Machine] | None,
           expected_modeling_rules_to_test: Iterable[str | Path] | None,
-          collector_class_args: tuple[Any, ...] = ()):
+          collector_class_args: tuple[Any, ...] = (),
+          disable_collect_packs_diff_master_bucket=True):
```
Comment (dorschw) [inline Tests/scripts/infrastructure_tests/test_collect_tests.py:132]: add to docstrings, explain why we disable it only sometimes

**Review by dorschw — COMMENTED** [review]
Very nice

File: Tests/scripts/collect_tests/collect_tests.py [hunk 3]
```diff
@@ -809,11 +811,42 @@ def _collect(self) -> CollectionResult | None:
...
+    def _collect_packs_diff_master_bucket(self) -> CollectionResult | None:
+
+        collected_packs: list[CollectionResult | None] = []
+
+        # diff between master and the last upload
+        collect_from = self._get_git_diff(upload_delta_from_last_upload=True)
+
+        for file_path in collect_from.changed_files:
+
+            path = PATHS.content_path / file_path
```
[the remark this answers is not in the document]
Comment (rshunim) [reply Tests/scripts/collect_tests/collect_tests.py:826]: Done

File: Tests/scripts/collect_tests/collect_tests.py [hunk 4]
```diff
@@ -808,11 +808,39 @@ def _collect(self) -> CollectionResult | None:
...
         return CollectionResult.union([
             self.__collect_from_changed_files(collect_from.changed_files),
-            self.__collect_packs_from_which_files_were_removed(collect_from.pack_ids_files_were_removed_from)
+            self.__collect_packs_from_which_files_were_removed(collect_from.pack_ids_files_were_removed_from),
+            changed_packs,
+            packs_with_removed_files
         ])
 
+    def __collect_packs_diff_master_bucket(self) -> tuple[CollectionResult | None, CollectionResult | None]:
+
```
[the remark this answers is not in the document]
Comment (rshunim) [reply Tests/scripts/collect_tests/collect_tests.py:820]: Done

File: Tests/scripts/collect_tests/collect_tests.py [hunk 5]
```diff
@@ -809,11 +811,42 @@ def _collect(self) -> CollectionResult | None:
...
+            except NothingToCollectException as e:
+                logger.info(e.message)
+            except Exception as e:
+                logger.exception(f'Error while collecting pack for {path}', exc_info=True, stack_info=True)
+                raise e
+
+        # union with collected_packs since changed_files and since files were removed
+        return CollectionResult.union(tuple(itertools.chain(collected_packs,
+                                                            [self._collect_packs_from_which_files_were_removed
+                                                             (collect_from.pack_ids_files_were_removed_from)])))
```
[the remark this answers is not in the document]
Comment (rshunim) [reply Tests/scripts/collect_tests/collect_tests.py:845]: Done

File: Tests/scripts/collect_tests/collect_tests.py [hunk 6]
```diff
@@ -809,11 +811,42 @@ def _collect(self) -> CollectionResult | None:
...
+                    reason_description=file_path,
+                    only_to_install=False
+                ))
+            except NothingToCollectException as e:
+                logger.info(e.message)
+            except Exception as e:
+                logger.exception(f'Error while collecting pack for {path}', exc_info=True, stack_info=True)
+                raise e
+
+        # union with collected_packs since changed_files and since files were removed
```
[the remark this answers is not in the document]
Comment (rshunim) [reply Tests/scripts/collect_tests/collect_tests.py:842]: Done

File: Tests/scripts/collect_tests/collect_tests.py (hunk 1 above)
Comment (rshunim) [reply Tests/scripts/collect_tests/collect_tests.py:1144]: Done

File: Tests/scripts/infrastructure_tests/test_collect_tests.py (hunk 2 above)
Comment (rshunim) [reply Tests/scripts/infrastructure_tests/test_collect_tests.py:132]: Done

**Review by dorschw — APPROVED** [review]

File: Tests/scripts/collect_tests/collect_tests.py [hunk 7]
```diff
@@ -1120,7 +1169,12 @@ def _get_git_diff(self) -> FilesToCollect:
 
         logger.debug(f'Getting changed files for {self.branch_name=}')
 
-        if os.getenv('IFRA_ENV_TYPE') == 'Bucket-Upload':
+        if upload_delta_from_last_upload:
+            logger.info('bucket upload: getting last commit from index')
+            previous_commit = get_last_commit_from_index(self.service_account, self.marketplace)
+            current_commit = 'origin/master'
```
Comment (dorschw) [inline Tests/scripts/collect_tests/collect_tests.py:1175]:
```suggestion
            current_commit = self.branch_name
```

**Pull request merged**